scikeras/_utils.py

```diff
@@ -1,60 +1,66 @@
 import inspect
 import os
 import random
-import warnings
 
+from contextlib import contextmanager
 from types import FunctionType
-from typing import Any, Callable, Dict, Iterable, Mapping, Sequence, Type, Union
+from typing import (
+    Any,
+    Callable,
+    Dict,
+    Generator,
+    Iterable,
+    Mapping,
+    Sequence,
+    Type,
+    Union,
+)
 
 import numpy as np
 import tensorflow as tf
 
-from numpy.lib.arraysetops import isin
 from tensorflow.keras import losses as losses_mod
 from tensorflow.keras import metrics as metrics_mod
 from tensorflow.keras import optimizers as optimizers_mod
+from tensorflow.python.eager import context
+from tensorflow.python.framework import config, ops
 
 
 DIGITS = frozenset(str(i) for i in range(10))
 
 
-class TFRandomState:
-    def __init__(self, seed):
-        self.seed = seed
-        self._not_found = object()
+@contextmanager
+def tensorflow_random_state(seed: int) -> Generator[None, None, None]:
+    # Save values
+    origin_gpu_det = os.environ.get("TF_DETERMINISTIC_OPS", None)
+    orig_random_state = random.getstate()
+    orig_np_random_state = np.random.get_state()
+    if context.executing_eagerly():
+        tf_random_seed = context.global_seed()
+    else:
+        tf_random_seed = ops.get_default_graph().seed
 
-    def __enter__(self):
-        warnings.warn(
-            "Setting the random state for TF involves "
-            "irreversibly re-setting the random seed. "
```

Comment:
Why can this warning be removed?

Reply:
I believe the combination of seeing the global seed to None and config.disable_op_determinism() resolves the issue. But you are right to bring it up, I do want to double check

Reply:
Here's my quick test:

```python
from typing import List
import tensorflow as tf

from scikeras._utils import tensorflow_random_state

without_seed: List[float] = []
with_seed: List[float] = []

for _ in range(10):
    # without setting the random state we should get distinct random values
    # and having set the random state previously does not carry over
    without_seed.append(tf.random.uniform([1]).numpy()[0])
    without_seed.append(tf.random.uniform([1]).numpy()[0])
    # if we set the random state twice, we get the same value
    with tensorflow_random_state(42):
        with_seed.append(tf.random.uniform([1]).numpy()[0])
    with tensorflow_random_state(42):
        with_seed.append(tf.random.uniform([1]).numpy()[0])

assert len(without_seed) == len(set(without_seed))
assert len(set(with_seed)) == 1
```

So, at least for this branch, setting the random state has no persistent global effect as far as I can tell.

Reply:
That quick test should go in the tests?

Reply:
Well, not directly: it is testing a private method and is also using TensorFlow directly, which is not reflective of this packages intent. But we could lift some of it (namely doing a couple loops to ensure reversibility of the global state) into the [existing higher level test](https://github.com/adriangb/scikeras/blob/67de2388f34eecd1f09d4b1deaeb8ada5c60dc1b/tests/test_parameters.py#L31) (and we probably want to remove [this test](https://github.com/adriangb/scikeras/blob/67de2388f34eecd1f09d4b1deaeb8ada5c60dc1b/tests/test_parameters.py#L65), which is just testing implementation details).

Does that sound reasonable to you?

Reply:
:+1: for not testing private functions.

Why shouldn't `tf_random_state` be public?

Reply:
No particular reason, just expanding the API. I'm more open to it now that it looks more reliable + it's simplified into to a function based CM instead of a class.

Reply:
Let's move this conversation to #253.